controllers/machineset_controller.go

```diff
@@ -273,6 +273,9 @@ func (r *MachineSetReconciler) syncReplicas(ms *clusterv1.MachineSet, machines [
 			if machine.Spec.Bootstrap.ConfigRef != nil {
 				bootstrapConfig, err = external.CloneTemplate(r.Client, machine.Spec.Bootstrap.ConfigRef, machine.Namespace)
 				if err != nil {
+					if err := r.Client.Delete(context.TODO(), infraConfig); !apierrors.IsNotFound(err) {
+						klog.Errorf("Failed to cleanup infrastructure configuration object after bootstrap clone error: %v", err)
+					}
```

Comment:
Instead of attempting to delete here, would it make sense to move the cloning of the bootstrap template before the cloning of the infrastructure template?

Reply:
+1

Reply:
> @vincepri added a commit to change the ordering and rebased to account for the addition of a context to `external.CloneTemplate`

@detiber  Will this cause the leak of bootstrapconfig?